src/Compilers/Core/Portable/DiagnosticAnalyzer/AnalysisScope.cs

```diff
@@ -45,31 +46,65 @@ internal class AnalysisScope
+        /// <summary>
+        /// Flag indicating if this is a partial analysis for the corresponding <see cref="CompilationWithAnalyzers"/>,
+        /// i.e. <see cref="IsTreeAnalysis"/> is true and/or <see cref="Analyzers"/> is a subset of <see cref="CompilationWithAnalyzers.Analyzers"/>.
+        /// </summary>
+        public bool IsPartialAnalysis { get; }
+
+        public AnalysisScope(Compilation compilation, ImmutableArray<DiagnosticAnalyzer> analyzers, bool hasAllAnalyzers, bool concurrentAnalysis, bool categorizeDiagnostics)
+            : this(compilation.SyntaxTrees, analyzers, isPartialAnalysis: !hasAllAnalyzers, filterTreeOpt: null, filterSpanOpt: null, isSyntaxOnlyTreeAnalysis: false, concurrentAnalysis: concurrentAnalysis, categorizeDiagnostics: categorizeDiagnostics)
         {
         }
 
         public AnalysisScope(ImmutableArray<DiagnosticAnalyzer> analyzers, SyntaxTree filterTree, TextSpan? filterSpan, bool syntaxAnalysis, bool concurrentAnalysis, bool categorizeDiagnostics)
-            : this(SpecializedCollections.SingletonEnumerable(filterTree), analyzers, filterTree, filterSpan, syntaxAnalysis, concurrentAnalysis, categorizeDiagnostics)
+            : this(SpecializedCollections.SingletonEnumerable(filterTree), analyzers, isPartialAnalysis: true, filterTree, filterSpan, syntaxAnalysis, concurrentAnalysis, categorizeDiagnostics)
         {
             Debug.Assert(filterTree != null);
         }
 
-        private AnalysisScope(IEnumerable<SyntaxTree> trees, ImmutableArray<DiagnosticAnalyzer> analyzers, SyntaxTree filterTreeOpt, TextSpan? filterSpanOpt, bool isSyntaxOnlyTreeAnalysis, bool concurrentAnalysis, bool categorizeDiagnostics)
+        private AnalysisScope(IEnumerable<SyntaxTree> trees, ImmutableArray<DiagnosticAnalyzer> analyzers, bool isPartialAnalysis, SyntaxTree filterTreeOpt, TextSpan? filterSpanOpt, bool isSyntaxOnlyTreeAnalysis, bool concurrentAnalysis, bool categorizeDiagnostics)
         {
+            Debug.Assert(isPartialAnalysis || FilterTreeOpt == null);
+            Debug.Assert(isPartialAnalysis || FilterSpanOpt == null);
+            Debug.Assert(isPartialAnalysis || !isSyntaxOnlyTreeAnalysis);
+
             SyntaxTrees = trees;
             Analyzers = analyzers;
+            IsPartialAnalysis = isPartialAnalysis;
             FilterTreeOpt = filterTreeOpt;
             FilterSpanOpt = filterSpanOpt;
             IsSyntaxOnlyTreeAnalysis = isSyntaxOnlyTreeAnalysis;
             ConcurrentAnalysis = concurrentAnalysis;
             CategorizeDiagnostics = categorizeDiagnostics;
         }
 
-        public AnalysisScope WithAnalyzers(ImmutableArray<DiagnosticAnalyzer> analyzers)
+        private ImmutableHashSet<DiagnosticAnalyzer> _lazyAnalyzersSet;
+        public bool Contains(DiagnosticAnalyzer analyzer)
+        {
+            if (!IsPartialAnalysis)
+            {
+                Debug.Assert(containsCore());
```

Comment:
>containsCore() [](start = 29, length = 14)

This has side-effects but is only executed in DEBUG builds. Can we assert `_lazyAnalyzersSet != null` before calling `containsCore()`?

Reply:
Let me refactor the code so `_lazyAnalyzersSet` is define as a DEBUG only field, which should hopefully improve readability.